python_modules/dagster/dagster/core/definitions/pipeline_definition.py

```diff
@@ -1007,6 +1008,42 @@ def _checked_input_resource_reqs_for_mode(
     return resource_reqs
 
 
+def _is_input_resolvable(graph_def, input_def, node, upstream_nodes):
+    # If input is not loadable via config, check if loadable via top-level input (meaning it is mapped all the way up the graph composition).
+    if (
+        not input_def.dagster_type.loader
+        and not input_def.dagster_type.kind == DagsterTypeKind.NOTHING
+        and not input_def.root_manager_key
+        and not input_def.has_default_value
+    ):
+        return _is_input_resolved_from_top_level(graph_def, input_def, node, upstream_nodes)
+    else:
+        return True
+
+
+def _is_input_resolved_from_top_level(graph_def, input_def, node, upstream_nodes):
+    from dagster.core.definitions.input import InputPointer
+
+    input_name = input_def.name
+    node_name = node.name
+
+    upstream_nodes = upstream_nodes[::-1]
+    for upstream_node in upstream_nodes:
```

Comment:
nit: context on why reversing

Reply:
Upstream nodes are added in order of topological level, we explicitly want to check the nodes that directly contain the current node, which are the ones at the end of the list.

Reply:
will add a comment tho.